### ISSUE TITLE
Move file refactoring sometimes produces conflicting edits

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/sdk#55190]
This is from crash reporting:
```
Thread 0 [MAIN THREAD]
ConflictingEditException: ConflictingEditException: {"offset":8,"length":27,"replacement":"..."} conflicts with {"offset":8,"length":27,"replacement":"..."}
  at addEditForSource	(protocol_internal.dart:76)
  at SourceFileEdit.add	(protocol_common.dart:4420)
  at FileEditBuilderImpl._addEdit	(change_builder_core.dart:578)
  at FileEditBuilderImpl._addEditBuilder	(change_builder_core.dart:593)
  at FileEditBuilderImpl.addSimpleReplacement	(change_builder_core.dart:537)
  at MoveFileRefactoringImpl._appendChangeForFile.<fn>	(move_file.dart:144)
  at ChangeBuilderImpl.addDartFileEdit	(change_builder_core.dart:151)
  at MoveFileRefactoringImpl._appendChangeForFile	(move_file.dart:140)
  at <asynchronous gap>	(async)
  at MoveFileRefactoringImpl._appendChangesForResource	(move_file.dart:224)
  at <asynchronous gap>	(async)
  at MoveFileRefactoringImpl._appendChangesForResource	(move_file.dart:227)
  at <asynchronous gap>	(async)
  at MoveFileRefactoringImpl.createChange	(move_file.dart:88)
  at <asynchronous gap>	(async)
  at WillRenameFilesHandler._renameFile	(handler_will_rename_files.dart:53)
  at <asynchronous gap>	(async)
  at ServerStateMessageHandler.handleMessage	(handlers.dart:267)
  at <asynchronous gap>	(async)
  at LspAnalysisServer._handleRequestMessage	(lsp_analysis_server.dart:989)
  at <asynchronous gap>	(async)
  at LspAnalysisServer.handleMessage.<fn>.<fn>	(lsp_analysis_server.dart:454)
  at <asynchronous gap>	(async)
  at OperationPerformanceImpl.runAsync	(operation_performance.dart:172)
  at <asynchronous gap>	(async)
  at LspAnalysisServer.handleMessage.<fn>	(lsp_analysis_server.dart:437)
```
(the ellipses in the replacements are from removed text).

It appears that it's trying to replace the same piece of text with two different pieces of text. Hopefully there's enough information here to figure out why.

This might be related to https://github.com/dart-lang/sdk/issues/42710.

**Comment by bwilkerson** [comment]
@DanTup

**Comment by DanTup** [comment]
> the ellipses in the replacements are from removed text

Do you know what this text was? (my guess is that it's filenames from an `import`)

I suspect it's unrelated to https://github.com/dart-lang/sdk/issues/42710. I think that issue is from IntelliJ sending concurrent requests to move multiple files and one is cancelling another. Here, I think we're processing multiple file moves in a single operation but somehow computing different paths.

I thought the way we did this was by building a single big mapping of old/new paths which should result in only one possible edit for each, but it's been a while since I looked at the code. I'll see if I can find a repro.

Do you know if this issue is occurring a lot?

**Comment by bwilkerson** [comment]
> I think that issue is from IntelliJ ...

The stack trace indicates that it was running from an `LspAnalysisServer`, via the `WillRenameFilesHandler`. I don't think IntelliJ is involved.

**Comment by DanTup** [comment]
Actually, it looks like this code might be old:

```
 at MoveFileRefactoringImpl._appendChangesForResource	(move_file.dart:224)
```

It seems like `_appendChangesForResource` was removed in https://github.com/dart-lang/sdk/commit/09f3eb0c8b811a6753e01c3f8fff06caab7e1ae8 which was the change to support moving multiple items (which builds a table as I noted above). Prior to this change, we didn't support moving multiple items from the client, but if the provided item was a folder, we would enumerate the files within and compute changes.

My guess is that the old implementation didn't handle this loop properly (maybe sometimes we tried to edit a path multiple times in the loop) but maybe the new one does.

Based on the tags, I _think_ this change shipped in Dart 3.3, so I'm curious to know if this issue has only been showing up in 3.2 or if there are any cases in 3.3. If it's isolated to 3.2 then maybe there's nothing to do here. If it's happening in 3.3 then an updated stack trace may help.

**Comment by bwilkerson** [comment]
> Based on the tags, I think this change shipped in Dart 3.3 ...

If I'm reading the dashboard correctly, the most recent report was from 3.3.0-6.0.dev, so I think it's possible that this has been fixed.

**Comment by DanTup** [comment]
> If I'm reading the dashboard correctly, the most recent report was from 3.3.0-6.0.dev, so I think it's possible that this has been fixed.

The changes first tag is 3.3.0-11.0.dev so this sounds likely then (or at least, if the issue still exists it's likely to be significantly different based on the scale of changes in that commit).

I suggest we close this and re-file with an update stack trace if it comes up on >= 3.3.0-11.0.dev?
